### DISCUSSION TITLE
WebGL2 varying array equivalent in WebGPU

### DISCUSSION THREAD

**Discussion by shannon** [gpuweb/gpuweb#3290] [Q&A] [answered]
I am trying to implement a Cascaded Shadow Map solution in WebGPU but I have hit a roadblock that I am not sure how to work around. In WebGL2 I was able to do something like the following in my shader:

vertex:
```glsl
...

out vec4 v_ShadowTexcoords[6]; //<-- this is the important missing piece

void main() {
...
for (int i = 0; i < u_ShadowCount; ++i) {
        // u_ShadowMatrices are the precomputed projections matrices for the shadow maps
        v_ShadowTexcoords[i] = u_ShadowMatrices[i] * pos;
}
...
}
```

fragment:
```glsl

in vec4 v_ShadowTexcoords[6];

...

float offsetLookup(float shadowLayer, vec4 texcoord, vec2 offset)
{
    vec2 texelSize = vec2(textureSize(u_ShadowSamplers, 0).xy);
    texcoord.xyz /= texcoord.w;
    texcoord.xy += offset / texelSize;
    texcoord.w = texcoord.z;
    texcoord.z = shadowLayer;

    // u_ShadowSamplers is a 2d depth array texture with a layer for each directional light cascade or spot light shadow map
    return texture(u_ShadowSamplers, texcoord);
}
float shadowPCF(int shadowLayer, float bias) 
{
    vec4 texcoord   = v_ShadowTexcoords[shadowLayer]; //dynamically index the varying array
    texcoord.z -= bias;
    // Apply PCF
    float sum = 0.0; float x, y; 
    for (y = -1.5; y <= 1.5; y += 1.0)
    {
        for (x = -1.5; x <= 1.5; x += 1.0)
        {
            sum += offsetLookup(float(shadowLayer), texcoord, vec2(x, y)); 
        }
    }
    return sum / 16.0; 
}

```

This is just a snippet, but the key being that I could dynamically index the varying v_ShadowTexcoords to both set and read it from both shader stages. There seems to be no equivalent way to do this in wgsl because input/output locations are limited to 16 bytes. 

I have to do something like this in WebGPU:

``` wgsl
struct VertexOutput{
   @location(0) shadowTexcoords0: vec4<f32>,
   @location(1) shadowTexcoords1: vec4<f32>,
   @location(2) shadowTexcoords2: vec4<f32>,
   @location(3) shadowTexcoords3: vec4<f32>,
   @location(4) shadowTexcoords4: vec4<f32>,
   @location(5) shadowTexcoords5: vec4<f32>,
}
```

But then I can't dynamically index them to set or read them and I would need a bunch of branches. Or I do the shadow projection matrix math in the fragment shader but that seems like a bad idea. 

I also tried to do an array of pointers but that doesn't look possible either.

Is there a recommended way of handling this type of thing?

**Comment by alan-baker** [comment, marked as answer]
The problem is that each underlying API handles this a bit differently. Metal doesn't allow the declaration of arrays for pipeline IO and HLSL and GLSL will generate different mappings depending on the size of the data.

You can work around this problem by introducing some preamble/postamble to your shaders. For example:
```
struct VertexOutput{
   @location(0) shadowTexcoords0: vec4<f32>,
   @location(1) shadowTexcoords1: vec4<f32>,
   @location(2) shadowTexcoords2: vec4<f32>,
   @location(3) shadowTexcoords3: vec4<f32>,
   @location(4) shadowTexcoords4: vec4<f32>,
   @location(5) shadowTexcoords5: vec4<f32>,
   @builtin(position) pos : vec4<f32>,
}

@vertex
fn vert_main(...) -> VertexOutput {
  var outs : array<vec4<f32>, 6>;
  // Use dynamic indexing on outs to calculate the results.

  // Postlude
  return VertexOutput(outs[0], outs[1], outs[2], outs[3], outs[4], outs[5], pos);
}
```

The fragment shader would be similar. Also note that you could use a `var<private>` if you prefer not having to pass the array around to helper functions.
```
var<private> frag_ins : array<vec4<f32>, 6>;

@fragment
fn frag_main(ins : VertexOutput) -> ... {
  frag_ins[0] = ins.shadowTexcoords0;
  ...
  frag_ins[5] = ins.shadowTexcoords5;
  // Now use frag_ins everywhere
}
```

**Reply by shannon** [reply]
Ah ok, I think I can make that work. That's kind of what I was trying to do with the array of pointers but it didn't occur to me to just copy them over from another array, and  `VertexOutput(outs[0]...` seems like a clean way to do it. Thank you so much, I will try this out.
